var documenterSearchIndex = {"docs":
[{"location":"man/introduction/#About","page":"Introduction","title":"About","text":"","category":"section"},{"location":"man/introduction/","page":"Introduction","title":"Introduction","text":"This is a Julia implementation of option 29 of the Mendel software, originally programmed in Fortran. MendelKinship.jl contains new features, but its usage largely resembles Mendel by employing control files - requiring users to make a file to specify options instead of interacting with Julia directly. This is not the most Julian way of programming or usage experience... but unfortuantely we lack manpower to correct it. ","category":"page"},{"location":"man/introduction/#When-to-use-MendelKinship","page":"Introduction","title":"When to use MendelKinship","text":"","category":"section"},{"location":"man/introduction/","page":"Introduction","title":"Introduction","text":"MendelKinship.jl is capable of calculating the theoretical kinship coefficient Phi_ij as long as a valid pedigree structure is provided. When SNP markers are available, MendelKinship.jl can also calculate empirical kinship coefficients using GRM, robust GRM, or Method of Moments methods (see this paper and this paper for details). Here we recommend the Robust GRM or MoM (default) method because their estimates are more robust in the presence of rare alleles. ","category":"page"},{"location":"man/introduction/","page":"Introduction","title":"Introduction","text":"MendelKinship can optionally compare the empirical kinship and theoretical kinships to check suspect pedigree structures and reveal hidden relatedness. It can also reveal sample mixed ups or other laboratory errors that can lead to inaccurate empirical kinships.  The result is saved in a table sorted in descending order. We optioanlly output 2 interactive plots that allow users to quickly pinpoint pairs with the greatest theoretical vs empirical deviance. ","category":"page"},{"location":"man/introduction/#Using-PLINK-files-as-input","page":"Introduction","title":"Using PLINK files as input","text":"","category":"section"},{"location":"man/introduction/","page":"Introduction","title":"Introduction","text":"MendelKinship accepts PLINK binary format as input, in which case the triplets (data.bim, data.bed, data.fam) must all be present. One can import the data by specifying the following in the control file:","category":"page"},{"location":"man/introduction/","page":"Introduction","title":"Introduction","text":"plink_input_basename = data ","category":"page"},{"location":"man/introduction/","page":"Introduction","title":"Introduction","text":"However, sometimes the .fam file contains non-unique person id (2nd column of .fam file) across different pedigrees, which is currently not permitted in MendelKinship. A person's id cannot be repeated in other pedigrees, even if it is contextually clear that they are different persons.","category":"page"},{"location":"man/KinshipTutorial/#Tutorial-of-MendelKinship","page":"Tutorial","title":"Tutorial of MendelKinship","text":"","category":"section"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"versioninfo() # system info for reproducibility","category":"page"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"Julia Version 1.5.0\nCommit 96786e22cc (2020-08-01 23:44 UTC)\nPlatform Info:\n  OS: macOS (x86_64-apple-darwin18.7.0)\n  CPU: Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz\n  WORD_SIZE: 64\n  LIBM: libopenlibm\n  LLVM: libLLVM-9.0.1 (ORCJIT, skylake)","category":"page"},{"location":"man/KinshipTutorial/#Examples-data","page":"Tutorial","title":"Examples data","text":"","category":"section"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"Input data for this tutorial can be obtained on our Github, which were originally derived from the 1000 genome project. They contain 85 people and 253141 SNPs, half of which have maf 005. Using these founders' genotype, we simulated 127 extra people, resulting in 27 pedigrees and 212 people. Although the 85 individuals are treated as founders, they were actually somewhat related, and this is reflected in the kinship comparison in the 2nd example below. For more information on this dataset, please see Mendel's documentation example 29.4. ","category":"page"},{"location":"man/KinshipTutorial/#Example-1:-Theoretical-Kinship-Coefficient-Calculation","page":"Tutorial","title":"Example 1: Theoretical Kinship Coefficient Calculation","text":"","category":"section"},{"location":"man/KinshipTutorial/#Step-1:-Preparing-the-pedigree-files:","page":"Tutorial","title":"Step 1: Preparing the pedigree files:","text":"","category":"section"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"Recall what is a valid pedigree structure. Note that we require a header line. The extension .in have no particular meaning. Let's examine (the first few lines of) such an example:","category":"page"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":";head -10 \"Ped29a.in\"","category":"page"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"Pedigree,Person,Mother,Father,Sex,,,simTrait\n  1       ,  16      ,          ,          ,  F       ,          ,  29.20564,\n  1       ,  8228    ,          ,          ,  F       ,          ,  31.80179,\n  1       ,  17008   ,          ,          ,  M       ,          ,  37.82143,\n  1       ,  9218    ,  17008   ,  16      ,  M       ,          ,  35.08036,\n  1       ,  3226    ,  9218    ,  8228    ,  F       ,          ,  28.32902,\n  2       ,  29      ,          ,          ,  F       ,          ,  36.17929,\n  2       ,  2294    ,          ,          ,  M       ,          ,  42.88099,\n  2       ,  3416    ,          ,          ,  M       ,          ,  40.98316,\n  2       ,  17893   ,  2294    ,  29      ,  F       ,          ,  35.55038,","category":"page"},{"location":"man/KinshipTutorial/#Step-2:-Preparing-the-control-file","page":"Tutorial","title":"Step 2: Preparing the control file","text":"","category":"section"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"A control file gives specific instructions to MendelKinship. To perform theoretical kinship calculation, an minimal control file looks like the following:","category":"page"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":";cat \"control_just_theoretical_29a.txt\"","category":"page"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"#\n# Input and Output files.\n#\npedigree_file = Ped29a.in\n#\n# Analysis parameters for Kinship option.\n#\nkinship_output_file = just_theoretical_output.txt","category":"page"},{"location":"man/KinshipTutorial/#Step-3:-Run-the-analysis-in-Julia-REPL-or-directly-in-notebook","page":"Tutorial","title":"Step 3: Run the analysis in Julia REPL or directly in notebook","text":"","category":"section"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"# load packages\nusing MendelKinship, CSV\n\n# run analysis without printing OpenMendel's default messages\nKinship(\"control_just_theoretical_29a.txt\")\n\n# import result into dataframe for closer inspection\n# result = CSV.read(\"just_theoretical_output.txt\")","category":"page"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"     Welcome to OpenMendel's\n     Kinship analysis option\n \n \nReading the data.\n\nThe current working directory is \"/Users/biona001/.julia/dev/MendelKinship/docs/src/man\".\n\nKeywords modified by the user:\n\n  control_file = control_just_theoretical_29a.txt\n  kinship_output_file = just_theoretical_output.txt\n  pedigree_file = Ped29a.in\n \n \nAnalyzing the data.\n\n \n \nMendel's analysis is finished.","category":"page"},{"location":"man/KinshipTutorial/#Step-4:-Interpreting-the-result","page":"Tutorial","title":"Step 4: Interpreting the result","text":"","category":"section"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"MendelKinship should have generated the file just_theoretical_output.txt in your local directory. One can directly open the file, or import into the Julia environment for ease of manipulation using the DataFrames package. The fourth column contains the desired theoretical kinship coefficient. The 5th column contains the (deterministically) estimated Delta7 matrix. The 6th through the 14 columns contain the (stochastically) estimated Jacquard's 9 identity coefficients.","category":"page"},{"location":"man/KinshipTutorial/#Example-2:-Compare-theoretical/empirical-kinship-values","page":"Tutorial","title":"Example 2: Compare theoretical/empirical kinship values","text":"","category":"section"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"When both pedigree structure and complete SNP information are available, we can compare theoretical/empirical kinship coefficients. In practice, however, we often have individuals without genotype information, but nevertheless must be included in the pedigree structure. MendelKinship does not handle this situation yet, but an analysis option that supports these data is being developed. For now you can impute genotypes but keep in mind that the relationship comparison for these individuals who lack all genotype information will not be meaningful.   ","category":"page"},{"location":"man/KinshipTutorial/#Step-1:-Prepare-pedigree-file-and-SNP-data-file","page":"Tutorial","title":"Step 1: Prepare pedigree file and SNP data file","text":"","category":"section"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"The pedigree file is the same as the pedigree file in the previous example. The SNP definition file requires a header row, and should have approprietely placed commas. It may be informative to compare the following SNP definition file with the original \"SNP_def29a.in\" in Mendel Option 29a. ","category":"page"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":";head -10 \"SNP_def29a_converted.txt\"","category":"page"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"Locus,Chromosome,Basepairs,Allele1,Allele2\nrs3020701,19,90974,1,2\nrs56343121,19,91106,1,2\nrs143501051,19,93542,1,2\nrs56182540,19,95981,1,2\nrs7260412,19,105021,1,2\nrs11669393,19,107866,1,2\nrs181646587,19,107894,1,2\nrs8106297,19,107958,1,2\nrs8106302,19,107962,1,2","category":"page"},{"location":"man/KinshipTutorial/#Non-binary-PLINK-users","page":"Tutorial","title":"Non binary PLINK users","text":"","category":"section"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"The SNP data files in this case must be stored in PLINK BED file in SNP-major format, with an accompanying SNP definition file. For an explanation of what these are, see MendelBase documentation.","category":"page"},{"location":"man/KinshipTutorial/#Binary-PLINK-file-users","page":"Tutorial","title":"Binary PLINK file users","text":"","category":"section"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"If your have \"data.bim\", \"data.bed\", \"data.fam\" (i.e. the 3 triplet of PLINK files), then you can replace the 3 fields snpdata_file, snpdefinition_file, and pedigree_file in the next step with just 1 field:","category":"page"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"plink_input_basename = data.","category":"page"},{"location":"man/KinshipTutorial/#Step-2:-Preparing-control-file","page":"Tutorial","title":"Step 2: Preparing control file","text":"","category":"section"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"The following control file tells MendelKinship to compare theoretical kinship and empirical kinship, and output 2 interactive plots stored in .html format. ","category":"page"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":";cat \"control_compare_29a.txt\"","category":"page"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"#\n# Input and Output files.\n#\nsnpdata_file = SNP_data29a.bed\nsnpdefinition_file = SNP_def29a_converted.txt\npedigree_file = Ped29a.in\n#\n# Analysis parameters for Kinship option.\n#\ncompare_kinships = true\nkinship_plot = kinship_plot\nz_score_plot = z_score_plot","category":"page"},{"location":"man/KinshipTutorial/#Step-3:-Running-the-analysis","page":"Tutorial","title":"Step 3: Running the analysis","text":"","category":"section"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"# run analysis option without printing OpenMendel's default messages\nKinship(\"control_compare_29a.txt\")\n\n# import result into dataframe for closer inspection\n# result = CSV.read(\"Mendel_Output.txt\")","category":"page"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"     Welcome to OpenMendel's\n     Kinship analysis option\n \n \nReading the data.\n\nThe current working directory is \"/Users/biona001/.julia/dev/MendelKinship/docs/src/man\".\n\nKeywords modified by the user:\n\n  compare_kinships = true\n  control_file = control_compare_29a.txt\n  kinship_plot = kinship_plot\n  pedigree_file = Ped29a.in\n  snpdata_file = SNP_data29a.bed\n  snpdefinition_file = SNP_def29a_converted.txt\n  z_score_plot = z_score_plot\n \n \nAnalyzing the data.\n\nKinship plot saved.\nFisher's plot saved.\n \n \nMendel's analysis is finished.","category":"page"},{"location":"man/KinshipTutorial/#Step-4:-Interpreting-the-Result","page":"Tutorial","title":"Step 4: Interpreting the Result","text":"","category":"section"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"Founders which have 0 theoretical kinships often exhibit a non-zero empirical kinship. In the first row, person 26732 and 264 have 0 theoretical kinship but their empirical kinship is pretty close to 0.125 = 1/8. That is, these 2 people which we initially thought are unrelated, may be half siblings, grandparent-grandchild, or an avuncular pair. On the otherhand, the 8th row has a founder pair that has a -008 kinship (i.e. they are very unrelated), suggesting that the standard deviation of the moments estimator may have a wide spread. There may also have been a sample mix up. Another explanation is that we are only using one chromosome's worth of data and so the estimates of kinship may be imprecise. ","category":"page"},{"location":"man/KinshipTutorial/#Interactive-Plots-and-Tables","page":"Tutorial","title":"Interactive Plots and Tables","text":"","category":"section"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"MendelKinship automatically generates 2 figures and 1 table to allow the user to easily compare theoretical and empirical kinship, detect outliers, and observe skewnesses in distribution. Figures are saved in .html format to enable interactive sessions. To summarize, ","category":"page"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"The table containing all the pairwise kinship and theoretical comparisons is stored in kinship_file_output.txt. The table is sorted in descending order of the largest deviance between the theoretical and empiric kinship. The last column lists the Fisher's Z statistic (i.e. the number of standard deviations away from mean). \nThe 2 plots are stored in .html format, which should be automatically be generated in your directory. These figures can be examined interactively via jupyter notebook, as demonstrated below, or opened directly via the browser.","category":"page"},{"location":"man/KinshipTutorial/#Generated-Interactive-Plots-part-1:","page":"Tutorial","title":"Generated Interactive Plots part 1:","text":"","category":"section"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"The first interactive plot allows user to quickly identify which pairs of persons have an empirical kinship most deviated from their expected (theoretical) kinship. The midpoint is placed as an orange dot for interpretability. As an example, the first row in the table above is the highest point on the left most spread. Careful readers might observe that there is a wider spread on those with 0 expected theoretical kinship. This is expected, because most people are not related to each other, so we are making many more comparisons that have 0 expected kinship. ","category":"page"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"using PlotlyJS, CSV\n\n#import the files created from the previous example\nresult = CSV.read(\"kinship_file_output.txt\")\nname = Vector{String}(undef, size(result, 1))\n\n# label the data points according to the persons names\nfor i in 1:length(name)\n    name[i] = \"Person1=\" * string(result[i, 3]) * \", \" * \"Person2=\" * string(result[i, 4])\nend\n\n#create interactive graph\nfunction compare_kinship_plot()\n    trace1 = scatter(;x=result[:theoretical_kinship], \n        y=result[:empiric_kinship], mode=\"markers\", \n        name=\"empiric kinship\", text=name)\n    \n    trace2 = scatter(;x=[1/2, 1/4, 1/8, 1/16, 1/32, 1/64, 1/128, 0.0],\n        y=[1/2, 1/4, 1/8, 1/16, 1/32, 1/64, 1/128, 0.0], \n        mode=\"markers\", name=\"marker for midpoint\")\n        \n    layout = Layout(;title=\"Compare empiric vs theoretical kinship\",hovermode=\"closest\", \n        xaxis=attr(title=\"Theoretical kinship (θ)\", showgrid=false, zeroline=false),\n        yaxis=attr(title=\"Empiric Kinship\", zeroline=false))\n    \n    data = [trace1, trace2]\n    plot(data, layout)\nend\n\n# generate plot by this command\n# compare_kinship_plot()","category":"page"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"compare_kinship_plot (generic function with 1 method)","category":"page"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"If you call compare_kinship_plot(), you should get an interactive plot that looks like:","category":"page"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"(Image: svg)","category":"page"},{"location":"man/KinshipTutorial/#Generated-Interactive-Plots-part-2:","page":"Tutorial","title":"Generated Interactive Plots part 2:","text":"","category":"section"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"After comparing the theoretical and empirical kinships, as in the previous graph or through the outputted table directly, often one may wonder whether the observed differences between the two statistics are significantly different. As explained in our main OpenMendel paper (section 7), this difference can be precisely quantified by the Fisher's z transformation, which should give us samples from a standard normal distribution N(0 1). We ploted this statistic in plot 2, and at first glance, the distribution is approximately normal. In Julia, we can easily verify this by computing some summary statistics:","category":"page"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"function fishers_transform()\n    trace1 = histogram(x=result[:fishers_zscore], text=name)\n    data = [trace1]\n    \n    layout = Layout(barmode=\"overlay\", \n        title=\"Z-score plot for Fisher's statistic\",\n        xaxis=attr(title=\"Standard deviations\"),\n        yaxis=attr(title=\"count\"))\n    \n    plot(data, layout)\nend\n\n# generate plot by this command\n# fishers_transform()","category":"page"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"fishers_transform (generic function with 1 method)","category":"page"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"If you call fishers_transform(), you should get an interactive plot that looks like:","category":"page"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"(Image: svg)","category":"page"},{"location":"man/KinshipTutorial/#Compute-mean-and-variance","page":"Tutorial","title":"Compute mean and variance","text":"","category":"section"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"We can verify that the Fisher's statistic is approximately normal by checking its 1~4th moments:","category":"page"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"using Statistics, StatsBase\n\nmy_zscore = convert(Vector{Float64}, result[:fishers_zscore])\nmean(my_zscore), var(my_zscore)","category":"page"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"(-1.2084702388691578e-16, 1.0)","category":"page"},{"location":"man/KinshipTutorial/#Compute-skewness-and-excess-kurtosis","page":"Tutorial","title":"Compute skewness and excess kurtosis","text":"","category":"section"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"skewness(my_zscore), kurtosis(my_zscore)","category":"page"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"(0.0923976421835524, 0.10657222736224581)","category":"page"},{"location":"man/KinshipTutorial/#Conclusions","page":"Tutorial","title":"Conclusions","text":"","category":"section"},{"location":"man/KinshipTutorial/","page":"Tutorial","title":"Tutorial","text":"MendelKinship provides a rapid way to calculate the theoretical kinship, which requires accurate pedigrees. Calculation of empirical kinships is also possible if genotypes at multiple markers are availble. Further, it can compare these kinships using Fisher's Z statistic when both the pedigrees and markers are available. ","category":"page"},{"location":"#MendelKinship.jl","page":"Home","title":"MendelKinship.jl","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"Julia utilities for computing theoretical/empirical kinship and other identity coefficients","category":"page"},{"location":"","page":"Home","title":"Home","text":"MendelKinship.jl is a component of the umbrella OpenMendel project. ","category":"page"},{"location":"#Package-features","page":"Home","title":"Package features","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"Calculation of empirical and theoretical kinship coefficients, and their comparisons\nCalculation of Jacquard's 9 condensed identity coefficients\nCalculation of Delta7 matrix for non-inbred pedigrees","category":"page"},{"location":"","page":"Home","title":"Home","text":"For details, please read Chapter 5 of Mathematical and Statistical Methods for Genetic Analysis by Kenneth Lange, who wrote most of the code for this package. ","category":"page"},{"location":"#Installation","page":"Home","title":"Installation","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"Press ] to invoke the package manager mode and install the following:","category":"page"},{"location":"","page":"Home","title":"Home","text":"add https://github.com/OpenMendel/SnpArrays.jl\nadd https://github.com/OpenMendel/MendelSearch.jl\nadd https://github.com/OpenMendel/MendelBase.jl\nadd https://github.com/OpenMendel/MendelKinship.jl","category":"page"},{"location":"","page":"Home","title":"Home","text":"This package supports Julia v1.0+.","category":"page"},{"location":"#Manual-Outline","page":"Home","title":"Manual Outline","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"Pages = [\n    \"man/introduction.md\",\n    \"man/KinshipTutorial.md\"\n    \"man/keywords.md\",\n    \"man/api.md\"\n]\nDepth = 2","category":"page"},{"location":"man/api/#API","page":"API","title":"API","text":"","category":"section"},{"location":"man/api/","page":"API","title":"API","text":"List of MendelKinship.jl's functions.","category":"page"},{"location":"man/api/#Index","page":"API","title":"Index","text":"","category":"section"},{"location":"man/api/","page":"API","title":"API","text":"Pages = [\"api.md\"]","category":"page"},{"location":"man/api/#Functions","page":"API","title":"Functions","text":"","category":"section"},{"location":"man/api/","page":"API","title":"API","text":"Kinship\ncompare_kinships\ntheoretical_coefficients\nkinship_matrix\ndelta7_matrix\njacquard_coefficients\ncotwin_extension!\nidentity_state\ncompute_fishers_z\nplot_fisher_z","category":"page"},{"location":"man/api/#MendelKinship.Kinship","page":"API","title":"MendelKinship.Kinship","text":"Kinship(control_file = \"\"; args...)\n\nThis is the wrapper function for the Kinship analysis option. Keyword options should be specified in the control file. \n\n\n\n\n\n","category":"function"},{"location":"man/keywords/#Analysis-keywords","page":"Keyword Options","title":"Analysis keywords","text":"","category":"section"},{"location":"man/keywords/","page":"Keyword Options","title":"Keyword Options","text":"Keyword Default Value Allowed value Description\nkinship_output_file KinshipOutputFile.txt true/false OpenMendel generated output file with table of kinship coefficients\nrepetitions 1 Integer Repetitions for sharing statistics\nxlinked_analysis false beelean Whether markers are on the X chromosome\ncompare_kinships false boolean Whether we want to compare theoretical vs empiric kinship\nkinship_plot \"\" User defined file name A user specified name for a plot comparing theoretical and empiric kinship value\nz_score_plot \"\" User defined file name A user specified name for a plot of fisher's z statistic.\ngrm_method MoM GRM, MoM, Robust Method used for empiric kinship calculation. Defaults to MoM, but user could choose the more common GRM or Robust GRM methods instead. (Warning: Based on our experience, Fisher's z score is very unreliable if the GRM method is used for rare (maf < 0.2) snps)\nmaf_threshold 0.01 Real number between 0 and 1 The minor allele frequency threshold for the GRM computation\ndeviant_pairs false Integer less than n(n+1)2 Number of top deviant pairs (theoretical vs empiric kinship) the user wants to keep","category":"page"},{"location":"man/keywords/#See-also","page":"Keyword Options","title":"See also","text":"","category":"section"},{"location":"man/keywords/","page":"Keyword Options","title":"Keyword Options","text":"MendelBase documentation for general keywords common to most analysis package\nOld MendelKinship documentation","category":"page"}]
}
