<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Tutorial · MendelKinship</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit">MendelKinship</span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><a class="tocitem" href="../introduction/">Introduction</a></li><li class="is-active"><a class="tocitem" href>Tutorial</a><ul class="internal"><li><a class="tocitem" href="#Examples-data"><span>Examples data</span></a></li><li><a class="tocitem" href="#Example-1:-Theoretical-Kinship-Coefficient-Calculation"><span>Example 1: Theoretical Kinship Coefficient Calculation</span></a></li><li><a class="tocitem" href="#Example-2:-Compare-theoretical/empirical-kinship-values"><span>Example 2: Compare theoretical/empirical kinship values</span></a></li><li><a class="tocitem" href="#Interactive-Plots-and-Tables"><span>Interactive Plots and Tables</span></a></li><li><a class="tocitem" href="#Conclusions"><span>Conclusions</span></a></li></ul></li><li><a class="tocitem" href="../keywords/">Keyword Options</a></li><li><a class="tocitem" href="../api/">API</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Tutorial</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Tutorial</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/OpenMendel/MendelKinship.jl/blob/master/docs/src/man/KinshipTutorial.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Tutorial-of-MendelKinship"><a class="docs-heading-anchor" href="#Tutorial-of-MendelKinship">Tutorial of MendelKinship</a><a id="Tutorial-of-MendelKinship-1"></a><a class="docs-heading-anchor-permalink" href="#Tutorial-of-MendelKinship" title="Permalink"></a></h1><pre><code class="language-julia">versioninfo() # system info for reproducibility</code></pre><pre><code class="language-none">Julia Version 1.5.0
Commit 96786e22cc (2020-08-01 23:44 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin18.7.0)
  CPU: Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, skylake)</code></pre><h2 id="Examples-data"><a class="docs-heading-anchor" href="#Examples-data">Examples data</a><a id="Examples-data-1"></a><a class="docs-heading-anchor-permalink" href="#Examples-data" title="Permalink"></a></h2><p>Input data for this tutorial can be obtained on <a href="https://github.com/OpenMendel/MendelKinship.jl/tree/master/data/documentation_data">our Github</a>, which were originally derived from the 1000 genome project. They contain 85 people and 253141 SNPs, half of which have maf<span>$&lt; 0.05$</span>. Using these founders&#39; genotype, we simulated 127 extra people, resulting in 27 pedigrees and 212 people. Although the 85 individuals are treated as founders, they were actually somewhat related, and this is reflected in the kinship comparison in the 2nd example below. For more information on this dataset, please see Mendel&#39;s documentation example 29.4. </p><h2 id="Example-1:-Theoretical-Kinship-Coefficient-Calculation"><a class="docs-heading-anchor" href="#Example-1:-Theoretical-Kinship-Coefficient-Calculation">Example 1: Theoretical Kinship Coefficient Calculation</a><a id="Example-1:-Theoretical-Kinship-Coefficient-Calculation-1"></a><a class="docs-heading-anchor-permalink" href="#Example-1:-Theoretical-Kinship-Coefficient-Calculation" title="Permalink"></a></h2><h3 id="Step-1:-Preparing-the-pedigree-files:"><a class="docs-heading-anchor" href="#Step-1:-Preparing-the-pedigree-files:">Step 1: Preparing the pedigree files:</a><a id="Step-1:-Preparing-the-pedigree-files:-1"></a><a class="docs-heading-anchor-permalink" href="#Step-1:-Preparing-the-pedigree-files:" title="Permalink"></a></h3><p>Recall what is a <a href="https://openmendel.github.io/MendelBase.jl/#pedigree-file">valid pedigree structure</a>. Note that we require a header line. The extension <code>.in</code> have no particular meaning. Let&#39;s examine (the first few lines of) such an example:</p><pre><code class="language-julia">;head -10 &quot;Ped29a.in&quot;</code></pre><pre><code class="language-none">Pedigree,Person,Mother,Father,Sex,,,simTrait
  1       ,  16      ,          ,          ,  F       ,          ,  29.20564,
  1       ,  8228    ,          ,          ,  F       ,          ,  31.80179,
  1       ,  17008   ,          ,          ,  M       ,          ,  37.82143,
  1       ,  9218    ,  17008   ,  16      ,  M       ,          ,  35.08036,
  1       ,  3226    ,  9218    ,  8228    ,  F       ,          ,  28.32902,
  2       ,  29      ,          ,          ,  F       ,          ,  36.17929,
  2       ,  2294    ,          ,          ,  M       ,          ,  42.88099,
  2       ,  3416    ,          ,          ,  M       ,          ,  40.98316,
  2       ,  17893   ,  2294    ,  29      ,  F       ,          ,  35.55038,</code></pre><h3 id="Step-2:-Preparing-the-control-file"><a class="docs-heading-anchor" href="#Step-2:-Preparing-the-control-file">Step 2: Preparing the control file</a><a id="Step-2:-Preparing-the-control-file-1"></a><a class="docs-heading-anchor-permalink" href="#Step-2:-Preparing-the-control-file" title="Permalink"></a></h3><p>A control file gives specific instructions to <code>MendelKinship</code>. To perform theoretical kinship calculation, an minimal control file looks like the following:</p><pre><code class="language-julia">;cat &quot;control_just_theoretical_29a.txt&quot;</code></pre><pre><code class="language-none">#
# Input and Output files.
#
pedigree_file = Ped29a.in
#
# Analysis parameters for Kinship option.
#
kinship_output_file = just_theoretical_output.txt</code></pre><h3 id="Step-3:-Run-the-analysis-in-Julia-REPL-or-directly-in-notebook"><a class="docs-heading-anchor" href="#Step-3:-Run-the-analysis-in-Julia-REPL-or-directly-in-notebook">Step 3: Run the analysis in Julia REPL or directly in notebook</a><a id="Step-3:-Run-the-analysis-in-Julia-REPL-or-directly-in-notebook-1"></a><a class="docs-heading-anchor-permalink" href="#Step-3:-Run-the-analysis-in-Julia-REPL-or-directly-in-notebook" title="Permalink"></a></h3><pre><code class="language-julia"># load packages
using MendelKinship, CSV

# run analysis without printing OpenMendel&#39;s default messages
Kinship(&quot;control_just_theoretical_29a.txt&quot;)

# import result into dataframe for closer inspection
# result = CSV.read(&quot;just_theoretical_output.txt&quot;)</code></pre><pre><code class="language-none">     Welcome to OpenMendel&#39;s
     Kinship analysis option
 
 
Reading the data.

The current working directory is &quot;/Users/biona001/.julia/dev/MendelKinship/docs/src/man&quot;.

Keywords modified by the user:

  control_file = control_just_theoretical_29a.txt
  kinship_output_file = just_theoretical_output.txt
  pedigree_file = Ped29a.in
 
 
Analyzing the data.

 
 
Mendel&#39;s analysis is finished.</code></pre><h3 id="Step-4:-Interpreting-the-result"><a class="docs-heading-anchor" href="#Step-4:-Interpreting-the-result">Step 4: Interpreting the result</a><a id="Step-4:-Interpreting-the-result-1"></a><a class="docs-heading-anchor-permalink" href="#Step-4:-Interpreting-the-result" title="Permalink"></a></h3><p><code>MendelKinship</code> should have generated the file <code>just_theoretical_output.txt</code> in your local directory. One can directly open the file, or import into the Julia environment for ease of manipulation using the DataFrames package. The fourth column contains the desired theoretical kinship coefficient. The 5th column contains the (deterministically) estimated Delta7 matrix. The 6th through the 14 columns contain the (stochastically) estimated Jacquard&#39;s 9 identity coefficients.</p><h2 id="Example-2:-Compare-theoretical/empirical-kinship-values"><a class="docs-heading-anchor" href="#Example-2:-Compare-theoretical/empirical-kinship-values">Example 2: Compare theoretical/empirical kinship values</a><a id="Example-2:-Compare-theoretical/empirical-kinship-values-1"></a><a class="docs-heading-anchor-permalink" href="#Example-2:-Compare-theoretical/empirical-kinship-values" title="Permalink"></a></h2><p>When both pedigree structure and <em>complete</em> SNP information are available, we can compare theoretical/empirical kinship coefficients. In practice, however, we often have individuals without genotype information, but nevertheless must be included in the pedigree structure. <code>MendelKinship</code> does not handle this situation yet, but an analysis option that supports these data is being developed. For now you can impute genotypes but keep in mind that the relationship comparison for these individuals who lack all genotype information will not be meaningful.   </p><h3 id="Step-1:-Prepare-pedigree-file-and-SNP-data-file"><a class="docs-heading-anchor" href="#Step-1:-Prepare-pedigree-file-and-SNP-data-file">Step 1: Prepare pedigree file and SNP data file</a><a id="Step-1:-Prepare-pedigree-file-and-SNP-data-file-1"></a><a class="docs-heading-anchor-permalink" href="#Step-1:-Prepare-pedigree-file-and-SNP-data-file" title="Permalink"></a></h3><p>The pedigree file is the same as the pedigree file in the previous example. The SNP definition file requires a header row, and should have approprietely placed commas. It may be informative to compare the following SNP definition file with the original &quot;SNP_def29a.in&quot; in Mendel Option 29a. </p><pre><code class="language-julia">;head -10 &quot;SNP_def29a_converted.txt&quot;</code></pre><pre><code class="language-none">Locus,Chromosome,Basepairs,Allele1,Allele2
rs3020701,19,90974,1,2
rs56343121,19,91106,1,2
rs143501051,19,93542,1,2
rs56182540,19,95981,1,2
rs7260412,19,105021,1,2
rs11669393,19,107866,1,2
rs181646587,19,107894,1,2
rs8106297,19,107958,1,2
rs8106302,19,107962,1,2</code></pre><h4 id="Non-binary-PLINK-users"><a class="docs-heading-anchor" href="#Non-binary-PLINK-users">Non binary PLINK users</a><a id="Non-binary-PLINK-users-1"></a><a class="docs-heading-anchor-permalink" href="#Non-binary-PLINK-users" title="Permalink"></a></h4><p>The SNP data files in this case must be stored in PLINK BED file in SNP-major format, with an accompanying SNP definition file. For an explanation of what these are, see <a href="https://openmendel.github.io/MendelBase.jl/">MendelBase documentation</a>.</p><h4 id="Binary-PLINK-file-users"><a class="docs-heading-anchor" href="#Binary-PLINK-file-users">Binary PLINK file users</a><a id="Binary-PLINK-file-users-1"></a><a class="docs-heading-anchor-permalink" href="#Binary-PLINK-file-users" title="Permalink"></a></h4><p>If your have &quot;data.bim&quot;, &quot;data.bed&quot;, &quot;data.fam&quot; (i.e. the 3 triplet of PLINK files), then you can replace the 3 fields <code>snpdata_file</code>, <code>snpdefinition_file</code>, and <code>pedigree_file</code> in the next step with just 1 field:</p><p><code>plink_input_basename = data</code>.</p><h3 id="Step-2:-Preparing-control-file"><a class="docs-heading-anchor" href="#Step-2:-Preparing-control-file">Step 2: Preparing control file</a><a id="Step-2:-Preparing-control-file-1"></a><a class="docs-heading-anchor-permalink" href="#Step-2:-Preparing-control-file" title="Permalink"></a></h3><p>The following control file tells MendelKinship to compare theoretical kinship and empirical kinship, and output 2 interactive plots stored in .html format. </p><pre><code class="language-julia">;cat &quot;control_compare_29a.txt&quot;</code></pre><pre><code class="language-none">#
# Input and Output files.
#
snpdata_file = SNP_data29a.bed
snpdefinition_file = SNP_def29a_converted.txt
pedigree_file = Ped29a.in
#
# Analysis parameters for Kinship option.
#
compare_kinships = true
kinship_plot = kinship_plot
z_score_plot = z_score_plot</code></pre><h3 id="Step-3:-Running-the-analysis"><a class="docs-heading-anchor" href="#Step-3:-Running-the-analysis">Step 3: Running the analysis</a><a id="Step-3:-Running-the-analysis-1"></a><a class="docs-heading-anchor-permalink" href="#Step-3:-Running-the-analysis" title="Permalink"></a></h3><pre><code class="language-julia"># run analysis option without printing OpenMendel&#39;s default messages
Kinship(&quot;control_compare_29a.txt&quot;)

# import result into dataframe for closer inspection
# result = CSV.read(&quot;Mendel_Output.txt&quot;)</code></pre><pre><code class="language-none">     Welcome to OpenMendel&#39;s
     Kinship analysis option
 
 
Reading the data.

The current working directory is &quot;/Users/biona001/.julia/dev/MendelKinship/docs/src/man&quot;.

Keywords modified by the user:

  compare_kinships = true
  control_file = control_compare_29a.txt
  kinship_plot = kinship_plot
  pedigree_file = Ped29a.in
  snpdata_file = SNP_data29a.bed
  snpdefinition_file = SNP_def29a_converted.txt
  z_score_plot = z_score_plot
 
 
Analyzing the data.

Kinship plot saved.
Fisher&#39;s plot saved.
 
 
Mendel&#39;s analysis is finished.</code></pre><h3 id="Step-4:-Interpreting-the-Result"><a class="docs-heading-anchor" href="#Step-4:-Interpreting-the-Result">Step 4: Interpreting the Result</a><a id="Step-4:-Interpreting-the-Result-1"></a><a class="docs-heading-anchor-permalink" href="#Step-4:-Interpreting-the-Result" title="Permalink"></a></h3><p>Founders which have 0 theoretical kinships often exhibit a non-zero empirical kinship. In the first row, person 26732 and 264 have 0 theoretical kinship but their empirical kinship is pretty close to 0.125 = 1/8. That is, these 2 people which we initially thought are unrelated, may be half siblings, grandparent-grandchild, or an avuncular pair. On the otherhand, the 8th row has a founder pair that has a <span>$-0.08$</span> kinship (i.e. they are very <em>un</em>related), suggesting that the standard deviation of the moments estimator may have a wide spread. There may also have been a sample mix up. Another explanation is that we are only using one chromosome&#39;s worth of data and so the estimates of kinship may be imprecise. </p><h2 id="Interactive-Plots-and-Tables"><a class="docs-heading-anchor" href="#Interactive-Plots-and-Tables">Interactive Plots and Tables</a><a id="Interactive-Plots-and-Tables-1"></a><a class="docs-heading-anchor-permalink" href="#Interactive-Plots-and-Tables" title="Permalink"></a></h2><p><code>MendelKinship</code> automatically generates 2 figures and 1 table to allow the user to easily compare theoretical and empirical kinship, detect outliers, and observe skewnesses in distribution. Figures are saved in <code>.html</code> format to enable interactive sessions. To summarize, </p><ul><li>The table containing all the pairwise kinship and theoretical comparisons is stored in <code>kinship_file_output.txt</code>. The table is sorted in descending order of the largest deviance between the theoretical and empiric kinship. The last column lists the <a href="https://en.wikipedia.org/wiki/Fisher_transformation">Fisher&#39;s Z statistic</a> (i.e. the number of standard deviations away from mean). </li><li>The 2 plots are stored in .html format, which should be automatically be generated in your directory. These figures can be examined interactively via jupyter notebook, as demonstrated below, or opened directly via the browser.</li></ul><h3 id="Generated-Interactive-Plots-part-1:"><a class="docs-heading-anchor" href="#Generated-Interactive-Plots-part-1:">Generated Interactive Plots part 1:</a><a id="Generated-Interactive-Plots-part-1:-1"></a><a class="docs-heading-anchor-permalink" href="#Generated-Interactive-Plots-part-1:" title="Permalink"></a></h3><p>The first interactive plot allows user to quickly identify which pairs of persons have an empirical kinship most deviated from their expected (theoretical) kinship. The midpoint is placed as an orange dot for interpretability. As an example, the first row in the table above is the highest point on the left most spread. Careful readers might observe that there is a wider spread on those with 0 expected theoretical kinship. This is expected, because most people are not related to each other, so we are making many more comparisons that have 0 expected kinship. </p><pre><code class="language-julia">using PlotlyJS, CSV

#import the files created from the previous example
result = CSV.read(&quot;kinship_file_output.txt&quot;)
name = Vector{String}(undef, size(result, 1))

# label the data points according to the persons names
for i in 1:length(name)
    name[i] = &quot;Person1=&quot; * string(result[i, 3]) * &quot;, &quot; * &quot;Person2=&quot; * string(result[i, 4])
end

#create interactive graph
function compare_kinship_plot()
    trace1 = scatter(;x=result[:theoretical_kinship], 
        y=result[:empiric_kinship], mode=&quot;markers&quot;, 
        name=&quot;empiric kinship&quot;, text=name)
    
    trace2 = scatter(;x=[1/2, 1/4, 1/8, 1/16, 1/32, 1/64, 1/128, 0.0],
        y=[1/2, 1/4, 1/8, 1/16, 1/32, 1/64, 1/128, 0.0], 
        mode=&quot;markers&quot;, name=&quot;marker for midpoint&quot;)
        
    layout = Layout(;title=&quot;Compare empiric vs theoretical kinship&quot;,hovermode=&quot;closest&quot;, 
        xaxis=attr(title=&quot;Theoretical kinship (θ)&quot;, showgrid=false, zeroline=false),
        yaxis=attr(title=&quot;Empiric Kinship&quot;, zeroline=false))
    
    data = [trace1, trace2]
    plot(data, layout)
end

# generate plot by this command
# compare_kinship_plot()</code></pre><pre><code class="language-none">compare_kinship_plot (generic function with 1 method)</code></pre><p>If you call <code>compare_kinship_plot()</code>, you should get an interactive plot that looks like:</p><p><img src="../output_21_0.svg" alt="svg"/></p><h3 id="Generated-Interactive-Plots-part-2:"><a class="docs-heading-anchor" href="#Generated-Interactive-Plots-part-2:">Generated Interactive Plots part 2:</a><a id="Generated-Interactive-Plots-part-2:-1"></a><a class="docs-heading-anchor-permalink" href="#Generated-Interactive-Plots-part-2:" title="Permalink"></a></h3><p>After comparing the theoretical and empirical kinships, as in the previous graph or through the outputted table directly, often one may wonder whether the observed differences between the two statistics are significantly different. As explained in our main OpenMendel paper (section 7), this difference can be precisely quantified by the Fisher&#39;s z transformation, which should give us samples from a standard normal distribution <span>$N(0, 1)$</span>. We ploted this statistic in plot 2, and at first glance, the distribution is approximately normal. In Julia, we can easily verify this by computing some summary statistics:</p><pre><code class="language-julia">function fishers_transform()
    trace1 = histogram(x=result[:fishers_zscore], text=name)
    data = [trace1]
    
    layout = Layout(barmode=&quot;overlay&quot;, 
        title=&quot;Z-score plot for Fisher&#39;s statistic&quot;,
        xaxis=attr(title=&quot;Standard deviations&quot;),
        yaxis=attr(title=&quot;count&quot;))
    
    plot(data, layout)
end

# generate plot by this command
# fishers_transform()</code></pre><pre><code class="language-none">fishers_transform (generic function with 1 method)</code></pre><p>If you call <code>fishers_transform()</code>, you should get an interactive plot that looks like:</p><p><img src="../output_24_0.svg" alt="svg"/></p><h3 id="Compute-mean-and-variance"><a class="docs-heading-anchor" href="#Compute-mean-and-variance">Compute mean and variance</a><a id="Compute-mean-and-variance-1"></a><a class="docs-heading-anchor-permalink" href="#Compute-mean-and-variance" title="Permalink"></a></h3><p>We can verify that the Fisher&#39;s statistic is approximately normal by checking its 1~4th moments:</p><pre><code class="language-julia">using Statistics, StatsBase

my_zscore = convert(Vector{Float64}, result[:fishers_zscore])
mean(my_zscore), var(my_zscore)</code></pre><pre><code class="language-none">(-1.2084702388691578e-16, 1.0)</code></pre><h3 id="Compute-skewness-and-excess-kurtosis"><a class="docs-heading-anchor" href="#Compute-skewness-and-excess-kurtosis">Compute skewness and excess kurtosis</a><a id="Compute-skewness-and-excess-kurtosis-1"></a><a class="docs-heading-anchor-permalink" href="#Compute-skewness-and-excess-kurtosis" title="Permalink"></a></h3><pre><code class="language-julia">skewness(my_zscore), kurtosis(my_zscore)</code></pre><pre><code class="language-none">(0.0923976421835524, 0.10657222736224581)</code></pre><h2 id="Conclusions"><a class="docs-heading-anchor" href="#Conclusions">Conclusions</a><a id="Conclusions-1"></a><a class="docs-heading-anchor-permalink" href="#Conclusions" title="Permalink"></a></h2><p>MendelKinship provides a rapid way to calculate the theoretical kinship, which requires accurate pedigrees. Calculation of empirical kinships is also possible if genotypes at multiple markers are availble. Further, it can compare these kinships using Fisher&#39;s Z statistic when both the pedigrees and markers are available. </p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../introduction/">« Introduction</a><a class="docs-footer-nextpage" href="../keywords/">Keyword Options »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Friday 4 September 2020 18:14">Friday 4 September 2020</span>. Using Julia version 1.5.1.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
